# darevskia-pericentromere-analysis: Citations

## [Nextflow](https://doi.org/10.1038/nbt.3820) 

v21.10.6

> Di Tommaso P., Chatzou M., Floden E.W., Barja P.P., Palumbo E., Notredame C. (2017). Nextflow enables reproducible computational workflows. Nat Biotechnol 35:316-319. PMID: 28398311.

## [nf-core](https://doi.org/10.1038/s41587-020-0439-x)

> Ewels P., Peltzer A., Fillinger S., Patel H., Alneberg J., Wilm A., Garcia M.U., Di Tommaso P. & Nahnsen S. (2020). The nf-core framework for community-curated bioinformatics pipelines. Nat Biotechnol 38:276-278. PMID: 32055031.

## Pipeline tools

- [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) v0.11.9 

- [Trimmomatic](https://doi.org/10.1093/bioinformatics/btu170) v0.39 
  > Bolger A.M., Lohse M., Usadel B. (2014). Trimmomatic: a flexible trimmer for Illumina sequence data. Bioinformatics 30:2114-2120. PMID: 24695404.

- [NCBI Magic-BLAST](https://doi.org/10.1186/s12859-019-2996-x) v1.6.0
  > Boratyn G.M., Thierry-Mieg J., Thierry-Mieg D., Busby B., Madden T.L. (2019). Magic-BLAST, an accurate RNA-seq aligner for long and short reads. BMC Bioinformatics 20:405. PMID: 31345161.

- [TAREAN (TAndem REpeat ANalyzer)](https://doi.org/10.1093/nar/gkx257) v0.3.8 
  > Novák P., Robledillo L.A., Koblížková A., Vrbová I., Neumann P., Macas J. (2017). TAREAN: a computational tool for identification and characterization of satellite DNA from unassembled short reads. Nucleic Acids Res. 45:e111–e111. PMID: 28402514.

- [QUAST](https://doi.org/10.1093/bioinformatics/btt086) v5.0.2 
  > Gurevich A., Saveliev V., Vyahhi N., Tesler G. (2013). QUAST: quality assessment tool for genome assemblies. Bioinformatics 29:1072–1075. PMID: 23422339.

- [TRF (Tandem Repeats Finder)](https://doi.org/10.1093/nar/27.2.573) v4.09
  > Benson G. (1999). Tandem repeats finder: a program to analyze DNA sequences. Nucleic Acids Res. 27:573–580. PMID: 9862982.

- [Bowtie2](https://doi.org/10.1038/nmeth.1923) v2.4.4
  > Langmead B., Salzberg S.L. (2012). Fast gapped-read alignment with Bowtie 2. Nat Methods 9:357–359. PMID: 22388286.

## R packages

- [R](https://www.R-project.org/) v3.6.3
  > R Core Team (2023). R: A language and environment for statistical computing. R Foundation for Statistical Computing, Vienna, Austria.

- [tidyr](https://cran.r-project.org/web/packages/tidyr/index.html) v1.2.0
  > Wickham H., Vaughan D., Girlich M. (2023). tidyr: Tidy Messy Data.

- [dplyr](https://cran.r-project.org/web/packages/dplyr/index.html) v1.0.9
  > Wickham H., François R., Henry L., Müller K., Vaughan D. (2023). dplyr: A Grammar of Data Manipulation.

- [stringr](https://cran.r-project.org/web/packages/stringr/index.html) v1.4.0
  > Wickham H. (2022). stringr: Simple, Consistent Wrappers for Common String Operations.

- [utils](https://www.R-project.org/) v3.6.3
  > R Core Team (2023). R: A language and environment for statistical computing. R Foundation for Statistical Computing, Vienna, Austria.

- [ggplot2](https://cran.r-project.org/web/packages/ggplot2/index.html) v3.3.6
  > Wickham H. (2016). Ggplot2. Springer Science+Business Media, LLC, New York, NY.

## Software packaging/containerisation tools

- [Anaconda](https://anaconda.com)
  > Anaconda Software Distribution. Computer software. Vers. 2-2.4.0. Anaconda, Nov. 2016. Web.

- [Bioconda](https://doi.org/10.1038/s41592-018-0046-7)
  > Grüning B., Dale R., Sjödin A., Chapman B.A., Rowe J., Tomkins-Tinch C.H., Valieris R., Köster J., The Bioconda Team (2018). Bioconda: sustainable and comprehensive software distribution for the life sciences. Nat Methods 15:475-476. PMID: 29967506.

- [Docker](https://www.docker.com/)

- [Singularity](https://doi.org/10.1371/journal.pone.0177459)
  > Kurtzer G.M., Sochat V., Bauer M.W. (2017). Singularity: Scientific containers for mobility of compute. PLoS One 12:e0177459. PMID: 28494014.
